### PR TITLE
Add Uint512 and use in PoS

### PR DESCRIPTION
**Pull request by azarovh**
I didn't duplicate all the tests from `Uint256` and checked only corner cases specific to `Uint512`. I think it should be enough because the same code is already tested for `Uint128` and `Uint256`

### CODE REVIEW

**Review by stanislav-tkach — APPROVED** [review]

File: common/src/uint/impls.rs [hunk 1]
```diff
@@ -1027,4 +1076,13 @@ mod tests {
         test_op_equality(10000000000000000000000000);
         test_op_equality(100000000000000000000000000);
     }
+
+    #[test]
+    pub fn uint512_from_uint256() {
```
Comment (TheQuantumPhysicist) [inline common/src/uint/impls.rs:1081]: Please add more aggressive testing. See how I tested 128 vs 256.

**Review by TheQuantumPhysicist — APPROVED** [review]

**Review by iljakuklic — APPROVED** [review]

File: common/src/uint/impls.rs (hunk 1 above)
Comment (azarovh) [reply common/src/uint/impls.rs]: I rewrote this test and an old one using proptest. Please take a look

**Review by muursh — APPROVED** [review]

**Review by TheQuantumPhysicist — APPROVED** [review]
